lib/complete-provider.coffee

```diff
@@ -53,11 +58,19 @@ class CompleteProvider extends Provider
     'then', 'where'
   ]
 
-  initialize: (@editorView, @manager) ->
+  constructor: (@editor, @manager) ->
+    if (@editor.haskellCompletionProvider && @editor.haskellCompletionProvider != this)
+      # shouldn't happen, but log just in case
+      console.log "weird: editor already has a haskell completion provider"
+      @editor.haskellCompletionProvider.dispose()
```

Comment:
Could you just subscribe provider to all open editors (or rather, buffers)? `Buffer.onDidSave` and `Buffer.onWillSave` pass an `Object {path: uri}` as a first argument to callbacks, so `@currentBuffer` is not really needed anyway.
onSaved could be rewritten like this, for example:
```coffeescript
onSaved: ({path}) =>
   fileName=path
   ...
```